ChronoKeeper/KeeperRecordingService.cpp

```diff
@@ -0,0 +1,100 @@
+#include <iostream>
+#include <margo.h>
+#include <thallium.hpp>
+#include <thallium/serialization/stl/string.hpp>
+
+namespace tl = thallium;
+
+
+#ifdef INNA 
+//INNA :  for now assume all identifiers are uint64_t , revisit later on
+
+namespace chronolog
+{
+
+//#INNA :  for now assume all identifiers are uint64_t , revisit later 
+typedef uint64_t StoryId;
+typedef uint64_t StorytellerId;
+
+struct Event { uint64_t time; std::string record; };
+
+class KeeperRecordingService : public tl::provider<KeeperRecordingService> 
+{
+
+    private:
+
+    void hello(const std::string& name) {
+        std::cout << "Hello, " << name << std::endl;
+    }
+
+    void record_event(tl::request const& request, 
+                      //const StorytellerId& teller_id, const  StoryId& story_id, const std::string& record) 
+                       StorytellerId teller_id,  StoryId story_id, const std::string& record) 
+    {
+        std::cout << "recording {"<< teller_id<<":"<<story_id<<":"<< record <<"}"<< std::endl;
+        request.respond( static_cast<int>(record.size()) );
+    }
+
+    int record_int_values( int teller_id, int story_id) //, int teller_id)
+    { 
+        std::cout << "recording int by value  {"<<teller_id <<":"<<story_id<<"}"<< std::endl;
+	return story_id;
+    }
+    public:
+
+    KeeperRecordingService(tl::engine& tl_engine, uint16_t service_provider_id=1)
+    : tl::provider<KeeperRecordingService>(tl_engine, service_provider_id) {
+        define("hello", &KeeperRecordingService::hello, tl::ignore_return_value());
+        define("record_event", &KeeperRecordingService::record_event);
+        define("record_int_values", &KeeperRecordingService::record_int_values );
+    }
+
+    ~KeeperRecordingService() {
+        std::cout<<"KeeperRecordingService::destructor"<<std::endl;
+        get_engine().wait_for_finalize();
+    }
+};
+
+
+}// namespace chronolog
+#endif
+
+#include "KeeperIdCard.h"
+#include "KeeperRecordingService.h"
+
+// uncomment to test this class as standalone process
+int main(int argc, char** argv) {
+
+    uint16_t provider_id = 22;
+
```

Comment:
Can you use the existing RPC infrastructure to initialize the keeper engine? You could do ChronoLog::Singleton<tl::engine>::GetInstance(protocol.c_str(),THALLIUM_SERVER_MODE,true,numStreams_); and that would return a thallium engine. I am hoping the Chronokeeper also sticks to the programming style followed by ChronoVisor and Client where we have process+threads, i.e every process is have >=1 service threads. You could start with a single-threaded implementation to check correctness.